data/TaskDAO.js

```diff
@@ -0,0 +1,70 @@
+const Task = require('./Task');
+// In-memory implementation
+class TaskDAO {
+    constructor() {
+        this.tasks = [];
+        this.maxID = 0;
+    }
+
+    find(criteria) {
+        if(!criteria || Object.keys(criteria).length === 0) return this.tasks;
+        else {
+            return this.tasks.filter(task => this.matchesCriteria(task, criteria));
+        }
+    }
+
+    findById(id) {
+        return this.tasks.filter(task => task.id == id);
+    }
+
+    newTask(props) {
+        if(!this.validateNewTaskProps(props)) return;
+        let taskID = ++this.maxID;
+        let task = new Task(taskID, props.title, props.course, props.deadline, props.completed, props.scheduledDate);
+        this.tasks.push(task);
+        return task;
+    }
+
+    updateTask(props) {
+        if(!this.validateUpdateTaskProps(props)) return;
+        let task = this.findById(props.id)[0];
```

Comment:
This should also be a `const` you can modify properties of something that is a const, just not reassign it. 

Another approach to this could be 

```javascript
return Object.assign(task, props);
```

but if you want to make sure only the specific list get set then you could conditionally add them specifically.

```javascript
return Object.assign(task, {
          ...props.title && { title: props.title },
          ...props.course && { course: props.course },
          ...props.deadline && { deadline: props.deadline },
          ...props.copmleted && { title: props.completed },
          ...props.scheduleDate && { scheduleDate: props.scheduleDate },
        });
```